File: ruby/sk3/machine.rb

class Machine
  attr_reader :name, :tasks, :queue, :doing

  class Stat
    attr_reader :task, :low_water_mark, :count

    def initialize(_task, _count, _low_water_mark)
      @task = _task
      @count = _count
      @low_water_mark = _low_water_mark
    end

    def name
      @task.name
    end

    def <=>(other)
      name <=> other.name
    end

  end

  def initialize(event_queue, machine_name)
    @event_queue = event_queue
    @name = machine_name
    @tasks = {}
    @queue = []
    @doing = nil
    @event = method(:finish_task)
  end

  def init_sim
    start_next_task
    unless @doing
      @idle_start = @event_queue.curtime
    end
  end

  def add_tasks(*add)
    add.each { |t| add_task(t) }
  end

  def add_task(task)
    if task.machine && task.machine != self
      fail "Already assigned to another machine"
    end
    @tasks[task] = 0
    task.machine = self
  end

  def enqueue(task)
    fail "Can't queue a task on the wrong machine" unless @tasks[task]
    return if ((task == @doing) || @queue.include?(task))
    @queue.push task
    was_idle = !@doing
    start_next_task
    log_idle_end if was_idle
  end

  private

  def log_active_end
    return if @doing
    @idle_start = @event_queue.curtime
    length = 0
    statistics = @tasks.keys.inject([]) { |stats,task|
      count = @tasks[task]
      stats << Stat.new(task, count, task.low_water_mark)
      @tasks[task] = 0
      task.reset_low_water_mark!
      length += task.batch_time * count
      stats
    }.sort
    @event_queue.log(:active_end, self, length, statistics)
  end

  def log_idle_end
    return unless @doing && @idle_start
    length = @event_queue.curtime - @idle_start
    @idle_start = nil
    @event_queue.log(:idle_end, self, length)
  end

  def start_next_task
    return if @doing
    @doing = queue.shift
    if @doing
      @doing.start_batch
      @tasks[@doing] += 1
      @event_queue.add_event(@doing.batch_time) do
        finish_task
      end
    end
  end

  def finish_task
    return unless @doing
    task = @doing
    @doing = nil
    task.finish_batch
    start_next_task
    log_active_end
  end

end

File: cpp/src/machine.cpp
#include <algorithm>
#include <numeric>
#include <sstream>
#include "machine.hpp"
#include "task.hpp"
#include "event_queue.hpp"

namespace SK3 {

using namespace std;

namespace {

template<class M> shared_ptr<Machine> factory(const Config::Machine &cfg,
    shared_ptr<EventQueue> eventQ, TaskMap &tasks) {
  Machine::TaskCounts taskCounts;
  for_each(cfg.tasks.begin(), cfg.tasks.end(),
      [&tasks,&taskCounts](const string &taskName) {
        taskCounts[tasks.at(taskName)] = 0L;
      });
  return make_shared<M>(eventQ, cfg.name, taskCounts);
}

static map<string, machine_factory_t> factories {
  { "Machine", factory<Machine> }
};

} // anon namespace

machine_factory_t Machine::factoryFor(const string &variant) {
  return factories.at(variant);
}

Machine::Machine(shared_ptr<EventQueue> _eventQ, const string &_name,
    const TaskCounts &_tasks):
  eventQ(_eventQ),
  machineName(_name),
  eventFunc(bind(&Machine::finishTask, ref(*this))),
  tasks(_tasks) { }

void Machine::startNextTask() {
  if (!idle || runQ.empty()) return;
  idle = false;
  shared_ptr<Task> doing = runQ.front().lock();
  doing->startBatch();
  ++tasks[doing];
  eventQ->add_event(doing->batch_time, eventFunc);
}

typedef EventQueue::LogRecord LogRecord;
typedef LogRecord::Type LogType;

void Machine::finishTask() {
  if (idle) return;
  shared_ptr<Task> doing = runQ.front().lock();
  runQ.pop_front();
  doing->finishBatch();
  idle = true;
  startNextTask();
  if (idle) {
    idleStart = eventQ->now();
    ostringstream stats;
    bool first = true;
    Time duration = accumulate(tasks.begin(), tasks.end(), 0,
        [&stats,&first](Time sum, pair<const shared_ptr<Task>, long> &val) {
          const long count = val.second;
          shared_ptr<Task> task = val.first;
          if (first) {
            first = false;
          } else {
            stats << " ";
          }
          stats << task->name() << ':' << count
            << ',' << task->resetLowWaterMark();
          val.second = 0;
          return sum + task->batch_time * count;
        });
    eventQ->log(
      LogRecord(LogType::LOG_ACTIVE_END, duration, stats.str()),
      *this);
  }
}

void Machine::enqueue(shared_ptr<Task> task) {
  if (tasks.find(task) == tasks.end()) {
    // can't queue a task not assigned to this machine
    ostringstream errMsg;
    errMsg << "task '" << task->name()
      << "' is not assigned to machine '" << machineName
      << "'";
    throw halt_simulation(eventQ->now(), errMsg.str());
  }
  if (find(runQ.begin(), runQ.end(), weak_ptr<Task>(task)) != runQ.end()) {
    // already queued
    return;
  }
  runQ.push_back(task);
  bool wasIdle = idle;
  startNextTask();
  if (wasIdle && !idle) {
    Time duration = eventQ->now() - idleStart;
    if (duration <= 0) return;
    eventQ->log(
      LogRecord(LogType::LOG_IDLE_END, duration),
      *this);
  }
}

void Machine::init_sim() {
  idle = true;
  runQ.clear();
  idleStart = eventQ->now();
}

const string &Machine::name() const { return machineName; }
const Quantity Machine::buffer() const { return 0; }

}

